Add GitHub actions

KTibow opened the pull request:
For the best result, enable GitHub Pages on the master branch.

nneonneo commented:
Neat! I’m a little more comfortable with some kind of CI setup rather than a custom GitHub Actions pipeline though - for one, I spotted that this requires credentials (or at least username/email) in order to upload the report, and CI setups can be setup to operate on PRs too. I do like the idea of having a proper test infrastructure - it will definitely make things more predictable and consistent.

KTibow commented:
Well, only to commit. You can easily make it use [artifacts](https://github.com/marketplace/actions/upload-artifact) instead. Also, this isn't custom; they have a standard setup for python apps you can see [here](https://github.com/actions/starter-workflows/blob/master/ci/python-app.yml). If it makes you more comfortable, I can make it more plain.

KTibow commented:
Okay, implemented that. I modified the standard setup so that you get an artifact. If you unzip it and open report.html in your browser, you should get a nice pytest report. But other than that, it's just the [standard](https://github.com/actions/starter-workflows/blob/master/ci/python-app.yml).

KTibow commented:
Also, I'd still recommend you enable GitHub Pages, because then you get [this](https://ktibow.github.io/ffsend/) nice README webpage.

KTibow commented:
@nneonneo does my latest update make merging this more appealing?

KTibow commented:
@nneonneo did you forget about this?

nneonneo commented:
When I used to use Travis CI, any pull request or push would trigger a CI build complete with a nice hosted report (no need to futz with a local HTML file). It made it really easy to see if a PR broke the build. Is there anything like that for GH actions?

Also apologies for any latency in replying - it’s the holidays so I’m less available than usual.

KTibow commented:
Well, the problem is that it's hard to do this without triggering an infinite loop. One way is to have a gh-pages branch, and only have the report trigger on the master branch. You still need a PAT, though. I'll start working on a new version.

nneonneo commented on .github/workflows/pythonapp.yml:
```diff
@@ -0,0 +1,45 @@
+        git add report/
+        git commit -m "Add report"
+        pat=$(printenv GH_PAT)
+        git push https://${pat}@github.com/KTibow/ffsend.git
```
Could this be rewritten to`git push https://${GH_PAT}@github.com/${{github.repository}}.git`? Seems like the `printenv` is unnecessary. (Don't know if `${{github.repository}}` would just work - might need an environment variable, but seems nicer than hard-coding the name of the repo!)

KTibow replied:
Yup, apparently it could.

nneonneo commented on .github/workflows/pythonapp.yml:
```diff
@@ -0,0 +1,45 @@
+  build:
+
+    runs-on: ubuntu-latest
+
```
The default `pythonpackage.yml` action uses a matrix build to test on 2.7, 3.5, 3.6, 3.7. I like that, and I'm wondering if there's a good way to keep that. I realize the reports might end up needing to be merged somehow - maybe there's a way to dump all the reports into the same place and push with a post-action?

It's important to me that this package is tested with Python 2.7; although 2.7 is officially EOL tomorrow, there's still a nontrivial amount of software targeting 2.7, and keeping compatibility is not that difficult yet. I will revisit this if/when supporting 2.7 becomes cumbersome or there's another compelling reason to go 3.x-only.

KTibow replied:
Oh! I was using pythonapp.yml, not pythonpackage.yml. I'll change that.

KTibow replied:
Testing in Python 3.5 has a high failure rate.
```
=================== 10 failed, 3 passed, 1 skipped in 12.17s ===================
```

nneonneo commented on .github/workflows/pythonapp.yml:
```diff
@@ -0,0 +1,45 @@
+        pytest --html=../report/index.html
+    - name: Upload report
+      env:
+        GH_PAT: ${{ secrets.GH_PAT }}
```
I'm a *little* leery of using a PAT here. A deploy key seems like it would be a safer choice since it would be repo-specific (as opposed to a PAT, which AFAICT gives r/w access to all my public repos). Something like this looks workable: https://github.community/t5/GitHub-Actions/Github-Actions-CI-how-to-use-store-deploy-key-to-download-from/m-p/30251/highlight/true#M351

KTibow replied:
Well then, what email address should I use?

nneonneo commented on .github/workflows/pythonapp.yml:
```diff
@@ -0,0 +1,45 @@
+      run: |
+        git checkout gh-pages
+        git config --global user.name "The Tester"
+        rm -r report || true
```
`rm -rf report` should work here (instead of needing the `|| true`)

KTibow replied:
✔️

KTibow commented:
Sorry, I don't really know my way around that much. If you want to edit it I made it so you can edit it now, or I'll do it soon.

KTibow commented:
Make sure GitHub pages is set to the gh-pages branch.

nneonneo commented:
I got it all to work, thanks! If you look at the workflow I tweaked it to support the SSH deploy key instead of the PAT, and I also added badges and badge links. Thanks so much for doing the heavy lifting on the actions :)
